packages/mongodb-access-adapter/datagrip-access-adapter/src/main/kotlin/com/mongodb/jbplugin/accessadapter/datagrip/adapter/DataGripMongoDbDriver.kt

```diff
@@ -0,0 +1,129 @@
+/**
+ * Represents a MongoDB driver interface that uses a DataGrip
+ * connection to query MongoDB.
+ */
+
+package com.mongodb.jbplugin.accessadapter.datagrip.adapter
+
+import com.google.gson.Gson
+import com.intellij.database.dataSource.DatabaseConnection
+import com.intellij.database.dataSource.DatabaseConnectionManager
+import com.intellij.database.dataSource.LocalDataSource
+import com.intellij.database.dataSource.connection.ConnectionRequestor
+import com.intellij.database.run.ConsoleRunConfiguration
+import com.intellij.openapi.project.Project
+import com.mongodb.jbplugin.accessadapter.MongoDbDriver
+import com.mongodb.jbplugin.accessadapter.Namespace
+import org.bson.Document
+
+import kotlin.reflect.KClass
+import kotlin.time.Duration
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeout
+
+/**
+ * The driver itself. Shouldn't be used directly, but through the
+ * DataGripBasedReadModelProvider.
+ *
+ * @see com.mongodb.jbplugin.accessadapter.datagrip.DataGripBasedReadModelProvider
+ *
+ * @param project
+ * @param dataSource
+ */
+internal class DataGripMongoDbDriver(
+    private val project: Project,
+    private val dataSource: LocalDataSource
+) : MongoDbDriver {
+    private val gson = Gson()
+
+    override suspend fun <T : Any> runCommand(
+        command: Document,
+        result: KClass<T>,
+        timeout: Duration
+    ): T = withContext(
+        Dispatchers.IO
+    ) {
+        runQuery(
+            """db.runCommand(${command.toJson()})""",
+            result,
+            timeout
+        )[0]
+    }
+
+    override suspend fun <T : Any> findOne(
+        namespace: Namespace,
+        query: Document,
+        options: Document,
+        result: KClass<T>,
+        timeout: Duration
+    ): T? = withContext(Dispatchers.IO) {
+        runQuery(
+            """db.getSiblingDB("${namespace.database}")
+                 .getCollection("${namespace.collection}")
+                 .findOne(${query.toJson()}, ${options.toJson()}) """.trimMargin(),
+            result,
+            timeout
+        ).getOrNull(0)
+    }
+
+    override suspend fun <T : Any> findAll(
+        namespace: Namespace,
+        query: Document,
+        result: KClass<T>,
+        limit: Int,
+        timeout: Duration
+    ) = withContext(Dispatchers.IO) {
+        runQuery(
+            """db.getSiblingDB("${namespace.database}")
+                 .getCollection("${namespace.collection}")
```

Comment:
Do you want to account for escaping here? (This might already be addressed by your updates locally, but just checking)

Otherwise some kind of TODO might be nice, even if this is still WIP :)

Reply:
I added some escaping now, thanks for pointing out! I was escaping only when converting a string to a namespace using the `.toNs` method, but not in the namespace constructor. Now I fixed it and it should be safe.